master/buildbot/www/hooks/bitbucket.py

```diff
@@ -12,24 +12,170 @@
+        repo = payload['repository']
+        project = repo['name']
+        push = payload['push']
+        payload['ref'] = push['changes'][0]["new"]["name"]
+        repo_url = "git@bitbucket.org:%s.git" % payload['repository']['full_name']  # TOFIX: Hard coded as clone URL is not available
+        changes = self._process_change(payload, repo, repo_url, project)
+        log.msg("Received %d changes from Bitbucket" % len(changes))
+        return changes, 'git'
+
+    def handle_pullrequest_updated(self, payload):
+        return self.handle_pullrequest(payload, 'updated_on')
+
+    def handle_pullrequest_created(self, payload):
+        return self.handle_pullrequest(payload, 'created_on')
+
+    def _process_change(self, payload, repo, repo_url, project):
+        """
+        Consumes the JSON as a python object and returns the changes.
+
+        :arguments:
+            payload
+                Python Object that represents the JSON sent by Bitbucket Webhook event.
+        """
+        changes = []
+
+        #  TOFIX: Code from github hook code, is this functionality needed?
+        # refname = payload['ref']
+        #
+        #  We only care about regular heads, i.e. branches
+        # match = re.match(r"^refs/heads/(.+)$", refname)
+        # print "match:", match
+        # if not match:
+        #     log.msg("Ignoring refname `%s': Not a branch" % refname)
+        #     return changes
+        #
+
+        for change_ in payload['push']['changes']:
+            if change_['new'] == None:
+                # Branch deleted
+                continue
+            branch = change_['new']['name']
+            when_timestamp = dateparse(change_['new']['target']['date'])
+            for commit in change_['commits']:
+                log.msg("New revision: %s" % commit['hash'][:8])
+
+                change = {
+                    'author': '%s <%s>' % (commit['author']['user']['display_name'], commit['author']['user']['username']),
+                    'comments': commit['message'],
+                    'revision': commit['hash'],
+                    'when_timestamp': when_timestamp,
+                    'branch': branch,
+                    'revlink': commit['links']['html']['href'],
+                    'repository': repo_url,  # The clone URL used to match in change filter
+                    'project': project,
+                    'properties': {}
+                }
+
+                if callable(self._codebase):
+                    change['codebase'] = self._codebase(payload)
+                elif self._codebase is not None:
+                    change['codebase'] = self._codebase
+
+                changes.append(change)
+
+        return changes
+
+    def handle_pullrequest(self, payload, timestamp_key):
+        changes = []
+        pr_number = payload["pullrequest"]['id']
+        refname = 'refs/pull/%d/merge' % (pr_number,)  # TOFIX: This is from github hook code, someone verify if this makes sense.
```

Comment:
This is a magic ref introduced by github,which always produce a mergecommit to the destination branch. Need to verify if bitbucket has similar feature.